File: Haskell/squirrelWarnsdroff.hs
{- Programma Haskell per risolvere il problema del giro del cavallo usando 
l'algoritmo di Warnsdorff-Squirrel. -}

import Data.List (sortBy)
{- La libreria Data.List è necessaria per utilizzare la funzione sortBy che ordina 
una lista in base a un criterio specifico. -}

import Data.Ord (comparing)
{- La libreria Data.Ord è necessaria per utilizzare la funzione comparing che crea 
un criterio di ordinamento basato su una funzione di proiezione. -}

import Data.Maybe (listToMaybe)
{- La libreria Data.Maybe è utilizzata per la funzione listToMaybe che converte una
 lista in un Maybe, prendendo il primo elemento della lista se esiste. -}

import Data.IORef (IORef, newIORef, readIORef, writeIORef)
{- La libreria Data.IORef è utilizzata per creare e manipolare riferimenti mutabili 
(IORef) necessari per gestire lo stato mutabile nel programma. -}

import Text.Read (readMaybe)
{- La libreria Text.Read è necessaria per utilizzare la funzione readMaybe che tenta 
di leggere un valore da una stringa e restituisce Maybe. -}

import System.Timeout (timeout)
{- La libreria System.Timeout è necessaria per utilizzare la funzione timeout che 
termina l'esecuzione a seguito di un tempo prestabilito. -}

-- Coordinate del cavallo
type Posizione = (Int, Int)
-- Scacchiera che contiene le mosse che compie il cavallo
type Scacchiera = [[Int]]

-- Movimenti del cavallo
mosseCavallo :: [Posizione]
mosseCavallo = [(2, 1), (1, 2), (-1, 2), (-2, 1), (-2, -1), (-1, -2), (1, -2), (2, -1)]

main :: IO ()
main = do
    dimensione <- leggiDimensioneScacchiera
    posizioneIniziale <- leggiPosizione dimensione
    putStrLn "Attendere la soluzione..."
    risultato <- timeout (120 * 1000000) (risolviGiroCavallo dimensione posizioneIniziale)
    case risultato of
        Just (Just soluzione) -> stampaScacchiera soluzione
        Just Nothing -> putStrLn "Soluzione non trovata."
        Nothing -> putStrLn "Tempo scaduto. Non è stata trovata una soluzione entro il tempo limite."

{- La funzione leggiDimensioneScacchiera legge la dimensione della scacchiera da tastiera
   e verifica la validità:
    - restituisce un intero che rappresenta la dimensione della scacchiera. -}

leggiDimensioneScacchiera :: IO Int
leggiDimensioneScacchiera = do
    putStrLn "Inserisci la dimensione della scacchiera (intero compreso tra 5 e 70):"
    datoIngresso <- getLine
    case readMaybe datoIngresso of
        Just dimensione
            | dimensione < 5 -> do
                putStrLn "Non esiste una soluzione per scacchiere di dimensioni inferiori a 5."
                leggiDimensioneScacchiera
            | dimensione > 70 -> do
                putStrLn "Con questo algoritmo non è possibile risolvere il problema per scacchiere di dimensioni superiori a 70."
                leggiDimensioneScacchiera
            | otherwise -> return dimensione
        Nothing -> do
            putStrLn "Dimensione non valida. Inserisci un numero intero."
            leggiDimensioneScacchiera

{- La funzione leggiPosizione legge una posizione da tastiera e verifica la validità 
   rispetto alla dimensione della scacchiera:
    - l'unico argomento è la dimensione della scacchiera 
    - restituisce una coppia di interi che rappresenta la posizione. -}

leggiPosizione :: Int -> IO (Int, Int)
leggiPosizione dimensione = do
    putStrLn "Inserisci la posizione di partenza del cavaliere in formato (X,Y)"
    putStrLn $ "(X e Y interi compresi tra 0 e " ++ show (dimensione - 1) ++ "):"
    datoIngresso <- getLine
    case reads datoIngresso of
        [(pos, "")] -> return pos
        _ -> do
            putStrLn "datoIngresso non valido. Inserisci di nuovo (esempio: (3,3)):"
            leggiPosizione dimensione

{- La funzione inizializzaScacchiera crea una scacchiera NxN inizializzata a -1:
    - il primo argomento è la dimensione della scacchiera. -}

inizializzaScacchiera :: Int -> Scacchiera
inizializzaScacchiera dimensione = replicate dimensione (replicate dimensione (-1))

{- La funzione mossaValida verifica se una posizione è valida:
    - il primo argomento è la dimensione della scacchiera;
    - il secondo argomento è la scacchiera;
    - il terzo argomento è la posizione da verificare. -}

mossaValida :: Int -> Scacchiera -> Posizione -> Bool
mossaValida dimensione scacchiera (x, y) =
    x >= 0 && x < dimensione &&
    y >= 0 && y < dimensione &&
    (scacchiera !! x !! y) == -1

{- La funzione calcolaAccessibilita calcola il numero di mosse valide successive da una data       
   posizione:
    - il primo argomento è la dimensione della scacchiera;
    - il secondo argomento è la scacchiera;
    - il terzo argomento è la posizione da cui calcolare l'accessibilità. -}

calcolaAccessibilita :: Int -> Scacchiera -> Posizione -> Int
calcolaAccessibilita dimensione scacchiera (x, y) =
    length $ filter (mossaValida dimensione scacchiera) [(x + dx, y + dy) | (dx, dy) <- mosseCavallo]

{- La funzione ordinaMosse ordina le mosse in base all'accessibilità e alla distanza dal centro:
    - il primo argomento è la dimensione della scacchiera;
    - il secondo argomento è la scacchiera;
    - il terzo argomento è la lista delle posizioni da ordinare. -}

ordinaMosse :: Int -> Scacchiera -> [Posizione] -> [(Int, Posizione)]
ordinaMosse dimensione scacchiera mosse =
    sortBy (comparing fst) $
    map (\pos ->
        (calcolaAccessibilita dimensione scacchiera pos, pos)
    ) mosse

{- La funzione aggiornaScacchiera aggiorna la scacchiera con la nuova mossa:
    - il primo argomento è la scacchiera;
    - il secondo argomento è la posizione della mossa;
    - il terzo argomento è il numero della mossa corrente. -}

aggiornaScacchiera :: Scacchiera -> Posizione -> Int -> Scacchiera
aggiornaScacchiera scacchiera (x, y) mossa =
    take x scacchiera ++
    [take y (scacchiera !! x) ++
     [mossa] ++
     drop (y + 1) (scacchiera !! x)] ++
    drop (x + 1) scacchiera


{- La funzione risolviGiroCavallo risolve il problema del giro del cavallo usando l'algoritmo di
   Warnsdorff-Squirrel:
    - il primo argomento è la dimensione della scacchiera;
    - il secondo argomento è la posizione di partenza del cavaliere. -}

risolviGiroCavallo :: Int -> Posizione -> IO (Maybe Scacchiera)
risolviGiroCavallo dimensione partenza = do
    riferimentoScacchiera <- newIORef (dimensione * dimensione, inizializzaScacchiera dimensione)
    algoritmoWarnsdorffSquirrel dimensione (inizializzaScacchiera dimensione) partenza 1 riferimentoScacchiera

{- La funzione algoritmoWarnsdorffSquirrel implementa l'algoritmo di Warnsdorff-Squirrel per 
   risolvere il giro del cavallo:
    - il primo argomento è la dimensione della scacchiera;
    - il secondo argomento è la scacchiera corrente;
    - il terzo argomento è la posizione corrente del cavallo;
    - il quarto argomento è il numero della mossa corrente;
    - il quinto argomento è un riferimento IORef contenente lo stato delle caselle mancanti e 
      la scacchiera. -}

algoritmoWarnsdorffSquirrel :: Int -> Scacchiera -> Posizione 
                            -> Int -> IORef (Int, Scacchiera) -> IO (Maybe Scacchiera)
algoritmoWarnsdorffSquirrel dimensione scacchiera posizione mossa riferimentoScacchiera = do
    let scacchieraAggiornata = aggiornaScacchiera scacchiera posizione mossa
    let caselleMancanti = dimensione * dimensione - mossa
    writeIORef riferimentoScacchiera (caselleMancanti, scacchieraAggiornata)
    if mossa == dimensione * dimensione
        then return (Just scacchieraAggiornata)
        else do
            let prossimeMosse = filter (mossaValida dimensione scacchieraAggiornata) [(fst posizione + dx, snd posizione + dy) | (dx, dy) <- mosseCavallo]
            if null prossimeMosse
                then return Nothing
                else do
                    let mosseOrdinate = ordinaMosse dimensione scacchieraAggiornata prossimeMosse
                    tentaMossa mosseOrdinate scacchieraAggiornata
  where
    tentaMossa [] _ = return Nothing
    tentaMossa ((_, prossimaPosizione):resto) scacchieraCorrente = do
        risultato <- algoritmoWarnsdorffSquirrel dimensione scacchieraCorrente prossimaPosizione (mossa + 1) riferimentoScacchiera
        case risultato of
            Just soluzione -> return (Just soluzione)
            Nothing -> tentaMossa resto scacchieraCorrente

{- La funzione stampaScacchiera stampa la scacchiera:
    - l'unico argomento è la scacchiera da stampare. -}

stampaScacchiera :: Scacchiera -> IO ()
stampaScacchiera scacchiera = mapM_ (putStrLn . unwords . map (pad . show)) scacchiera
  where pad s = replicate (3 - length s) ' ' ++ s

File: ProLog/squirrelWarnsdroff.pl
/* Programma Prolog per risolvere il problema del giro del cavallo usando l'algoritmo di Warnsdorff-Squirrel con ricorsione in coda. */

main :-
    leggi_dimensione_scacchiera(Dimensione),
    leggi_posizione(Dimensione, InizioX, InizioY),
    statistics(runtime, [TempoInizio|_]),  % Inizia a monitorare il tempo di esecuzione
    write('Attendere la soluzione...'), nl,
    (   risolvi(Dimensione, (InizioX, InizioY), TempoInizio, ScacchieraFinale)
    ->  stampa_scacchiera(ScacchieraFinale)
    ;   write('Soluzione non trovata.'), nl
    ).

/* Risolve il problema del giro del cavallo.
   - Dimensione: dimensione della scacchiera.
   - PosizioneIniziale: posizione iniziale del cavallo (InizioX, InizioY).
   - TempoInizio: tempo di inizio esecuzione.
   - ScacchieraFinale: la scacchiera finale con il percorso del cavallo.
*/
risolvi(Dimensione, PosizioneIniziale, TempoInizio, ScacchieraFinale) :-
    inizializza_scacchiera(Dimensione, Scacchiera),
    algoritmo_warnsdorff(Dimensione, Scacchiera, PosizioneIniziale, 1, TempoInizio, ScacchieraFinale).

/* Implementa l'algoritmo Warnsdorff-Squirrel con ricorsione in coda.
   - Dimensione: dimensione della scacchiera.
   - Scacchiera: scacchiera corrente.
   - (X, Y): posizione corrente del cavallo.
   - Mossa: numero della mossa corrente.
   - TempoInizio: tempo di inizio esecuzione.
   - ScacchieraFinale: la scacchiera finale con il percorso completato.
*/
algoritmo_warnsdorff(Dimensione, Scacchiera, (X, Y), Mossa, TempoInizio, ScacchieraFinale) :-
    statistics(runtime, [TempoAttuale|_]),  % Ottiene il tempo corrente
    TempoTrascorso is TempoAttuale - TempoInizio,
    (TempoTrascorso > 120000 ->
        format('Tempo limite superato: ~w ms\n', [TempoTrascorso]),
        halt  % Interrompe l'esecuzione se supera 120 secondi (120000 ms)
    ; aggiorna_scacchiera(Scacchiera, (X, Y), Mossa, ScacchieraAggiornata),
        (   Mossa =:= Dimensione * Dimensione  % Caso base: tutte le celle sono state visitate
        ->  ScacchieraFinale = ScacchieraAggiornata
        ;   findall((NX, NY), 
                    (mosse_cavallo((X, Y), (NX, NY)), 
                     mossa_valida(Dimensione, ScacchieraAggiornata, (NX, NY))
                    ), 
                    Mosse),
            ordina_mosse(Dimensione, ScacchieraAggiornata, Mosse, MosseOrdinate),
            member(ProssimaMossa, MosseOrdinate),
            algoritmo_warnsdorff(Dimensione, ScacchieraAggiornata, ProssimaMossa,
                                 Mossa + 1, TempoInizio, ScacchieraFinale)
        )
    ).
/* Legge e valida la dimensione della scacchiera. 
*/
leggi_dimensione_scacchiera(Dimensione) :-
    write('Inserisci la dimensione della scacchiera (intero compreso tra 5 e 70): '),
    read(Ingresso),
    (   integer(Ingresso), Ingresso >= 5, Ingresso =< 70 ->
        Dimensione = Ingresso
    ;   (integer(Ingresso) ->
            write('Dimensione non valida. Deve essere compresa tra 5 e 70.')
        ;   write('Dato inserito non valido. Inserisci un numero intero.')
        ), nl, leggi_dimensione_scacchiera(Dimensione)
    ).

/* Legge e valida la posizione iniziale del cavallo. 
*/
leggi_posizione(N, InizioX, InizioY) :-
    repeat,
    write('Inserisci la posizione di partenza del cavallo (X,Y)'),
    format(' (X e Y interi compresi tra 0 e ~d): ', [N-1]),
    catch(read_term(user_input, Ingresso, []), _, fallimento),
    (   analizza_ingresso(Ingresso, N, InizioX, InizioY)
    ->  !
    ;   fallimento).

/* Funzione di fallimento chiamata in caso di errore durante la lettura 
*/
fallimento :-
    write('Dato inserito non valido o errore di sintassi. Riprova.\n'),
    fail.

/* Analizza il dato in ingresso e verifica che sia nel formato corretto (X,Y). 
*/
analizza_ingresso((X,Y), N, X, Y) :-
    integer(X), integer(Y), X >= 0, X < N, Y >= 0, Y < N.
analizza_ingresso(_, _, _, _) :-
    fail.

/* Inizializza la scacchiera con valori di default (-1). 
*/
inizializza_scacchiera(N, Scacchiera) :-
    length(Scacchiera, N), % Crea una lista di N righe
    maplist(crea_riga(N), Scacchiera).

/* Crea una riga della scacchiera con valori di default (-1). 
*/
crea_riga(N, Riga) :-
    length(Riga, N), % Crea una riga di lunghezza N
    maplist(=( -1), Riga). % Inizializza tutte le celle a -1

/* Genera le mosse possibili del cavallo a partire da una posizione. 
*/
mosse_cavallo((X, Y), (NuovoX, NuovoY)) :-
    member((DeltaX, DeltaY), [(2, 1), (1, 2), (-1, 2), (-2, 1),
                              (-2, -1), (-1, -2), (1, -2), (2, -1)]),
    NuovoX is X + DeltaX,
    NuovoY is Y + DeltaY,
    NuovoX >= 0,
    NuovoY >= 0.

/* Controlla se una mossa è valida (ossia se è dentro i confini della scacchiera e la cella non è  
   stata visitata). 
*/
mossa_valida(N, Scacchiera, (X, Y)) :-
    X >= 0, X < N, Y >= 0, Y < N,
    nth0(X, Scacchiera, Riga),
    nth0(Y, Riga, -1).

/* Calcola l'accessibilità di una posizione (numero di mosse valide possibili). 
*/
calcola_accessibilita_con_posizione(N, Scacchiera, (X, Y), Grado-(X, Y)) :-
    findall((NuovoX, NuovoY), (mosse_cavallo((X, Y), (NuovoX, NuovoY)), 
             mossa_valida(N, Scacchiera, (NuovoX, NuovoY))), Mosse),
    length(Mosse, Grado).

/* Ordina le mosse in base all'accessibilità (secondo la regola di Warnsdorff). 
*/
ordina_mosse(N, Scacchiera, Mosse, MosseOrdinate) :-
    maplist(calcola_accessibilita_con_posizione(N, Scacchiera), Mosse, MosseConAccessibilita),
    keysort(MosseConAccessibilita, MosseOrdinateConAccessibilita),
    estrai_valori(MosseOrdinateConAccessibilita, MosseOrdinate).

/* Estrae i valori da una lista di coppie chiave-valore. 
*/
estrai_valori([], []).
estrai_valori([_-Val|Resto], [Val|ValoriResto]) :-
    estrai_valori(Resto, ValoriResto).

/* Aggiorna la scacchiera con una nuova mossa. 
*/
aggiorna_scacchiera(Scacchiera, (X, Y), Mossa, NuovaScacchiera) :-
    (X >= 0, Y >= 0 ->  
        nth0(X, Scacchiera, Riga),  
        aggiorna_lista(Riga, Y, Mossa, NuovaRiga),  
        aggiorna_lista(Scacchiera, X, NuovaRiga, NuovaScacchiera)
    ;   throw(error(domain_error(not_less_than_zero, (X, Y)), _))).

/* Aggiorna una lista con un nuovo elemento in una posizione specifica. 
*/
aggiorna_lista([_|Resto], 0, Elem, [Elem|Resto]) :- !.
aggiorna_lista([Testa|Resto], Indice, Elem, [Testa|NuovoResto]) :-
    (Indice > 0 ->  
        NuovoIndice is Indice - 1,
        aggiorna_lista(Resto, NuovoIndice, Elem, NuovoResto)
    ;   throw(error(domain_error(not_less_than_zero, Indice), _))).

/* Stampa la scacchiera. 
*/
stampa_scacchiera(Scacchiera) :-
    stampa_righe(Scacchiera).

/* Stampa tutte le righe della scacchiera. 
*/
stampa_righe([]).
stampa_righe([Riga|Resto]) :-
    stampa_riga(Riga),
    nl,
    stampa_righe(Resto).

/* Stampa una singola riga della scacchiera. 
*/
stampa_riga([]).
stampa_riga([Elemento|Resto]) :-
    format('~d ', [Elemento]),  % Stampa l'elemento con uno spazio
    stampa_riga(Resto).
